ml-agents/mlagents/trainers/components/bc/model.py

```diff
@@ -0,0 +1,101 @@
+import tensorflow as tf
+import numpy as np
+from mlagents.trainers.models import LearningModel
+
+
+class BCModel(object):
+    def __init__(
+        self,
+        policy_model: LearningModel,
+        learning_rate: float = 3e-4,
+        anneal_steps: int = 0,
+    ):
+        """
+        Tensorflow operations to perform Behavioral Cloning on a Policy model
+        :param policy_model: The policy of the learning algorithm
+        :param lr: The initial learning Rate for behavioral cloning
+        :param anneal_steps: Number of steps over which to anneal BC training
+        """
+        self.policy_model = policy_model
+        self.expert_visual_in = self.policy_model.visual_in
+        self.obs_in_expert = self.policy_model.vector_in
+        self.make_inputs()
+        self.create_loss(learning_rate, anneal_steps)
+
+    def make_inputs(self) -> None:
+        """
+        Creates the input layers for the discriminator
+        """
+        self.done_expert = tf.placeholder(shape=[None, 1], dtype=tf.float32)
+        self.done_policy = tf.placeholder(shape=[None, 1], dtype=tf.float32)
+
+        if self.policy_model.brain.vector_action_space_type == "continuous":
+            action_length = self.policy_model.act_size[0]
+            self.action_in_expert = tf.placeholder(
+                shape=[None, action_length], dtype=tf.float32
+            )
+            self.expert_action = tf.identity(self.action_in_expert)
+        else:
+            action_length = len(self.policy_model.act_size)
+            self.action_in_expert = tf.placeholder(
+                shape=[None, action_length], dtype=tf.int32
+            )
+            self.expert_action = tf.concat(
+                [
+                    tf.one_hot(
+                        self.action_in_expert[:, i], self.policy_model.act_size[i]
+                    )
+                    for i in range(len(self.policy_model.act_size))
+                ],
+                axis=1,
+            )
+
+    def create_loss(self, learning_rate: float, anneal_steps: int) -> None:
+        """
+        Creates the loss and update nodes for the GAIL reward generator
```

Comment:
This says GAIL reward generator, but the code is in the BC model?

Reply:
Yep - typo (fixed)